api/envoy/extensions/filters/http/api_key_auth/v3/api_key_auth.proto

```diff
@@ -17,38 +16,95 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
+//    credentials:
+//      entries:
+//      - key: real-key
+//        client: user
+//    key_sources:
+//      entries:
+//      - header: "X-API-KEY"
 //
-message APIKeyAuth {
-  // keys used to authenticate the client.
-  // It should be a map of clientID to apiKey.
-  // The clientID serves solely for identification purposes and isn't used for authentication.
-  config.core.v3.DataSource keys = 1 [(udpa.annotations.sensitive) = true];
+message ApiKeyAuth {
+  // The credentials that are used to authenticate the clients.
+  Credentials credentials = 1 [(udpa.annotations.sensitive) = true];
+
+  // The key sources to fetch the key from the coming request.
+  KeySources key_sources = 2;
+}
 
-  // The header name to fetch the key.
-  // If multiple values are present in the given header, the filter rejects the request.
-  // Only one of authentication_header, authentication_query, or authentication_cookie should be set.
-  string authentication_header = 2;
+// API key auth configuration of per route or per virtual host or per route configuration.
+message ApiKeyAuthPerRoute {
+  // Route specific APIKeyAuth configuration. This is optional and could be used to override the
+  // filter level ApiKeyAuth configuration **partly**.
+  //
+  // For example, if route specific ``authentication_header`` is set, then the filter will fetch
+  // the key from route specific ``authentication_header``. But the filter level ``keys`` will
+  // still be used. If route specific non-empty ``keys`` is set, then the filter will use the
+  // route specific ``keys`` to authenticate the client but the filter level ``authentication_header``
+  // will still be used. If both ``authentication_header`` and ``keys`` are set, then the filter
+  // will use the route specific ``authentication_header`` and ``keys`` to authenticate the client.
+  ApiKeyAuth override_config = 1;
 
-  // The query parameter name to fetch the key.
-  // Only one of authentication_header, authentication_query, or authentication_cookie should be set.
-  string authentication_query = 3;
+  // A list of client ids that are allowed to access the routes of the scope. If the list is empty,
+  // then all authenticated clients are allowed. This provides very limited but simple authorization.
+  // If more complex authorization is required, then use the :ref:`HTTP RBAC filter
+  // <config_http_filters_rbac>` instead.
+  repeated string allowed_clients = 2;
+}
+
+// Single credential entry that contains the API key and the related client id.
+message Credential {
+  // The value of the unique API key.
+  string key = 1 [(validate.rules).string = {min_len: 1}];
+
+  // The unique id or identity that used to identify the client or consumer.
+  string client = 2 [(validate.rules).string = {min_len: 1}];
+}
+
+// The credentials that are used to authenticate the clients.
+message Credentials {
+  // The list of credentials.
+  repeated Credential entries = 1 [(validate.rules).repeated = {min_items: 1}];
```

Comment:
Good point. I wraped them to a message because I want to implement the partly configuration overriding.

But I think I can do it by remove the `[(validate.rules).repeated = {min_items: 1}]` and make the non-zero list an explict overriding.

That have same effect but the API is much simpler. 

Thanks for the great point.